src/services/workflow.service.js

```diff
@@ -67,17 +69,154 @@ class WorkflowService extends GQuery {
+      fetchPolicy: 'no-cache'
+    }).then((response) => {
+      this.mutations = response.data.__schema.mutationType.fields
+      this.types = response.data.__schema.types
+      processMutations(this.mutations, this.types)
+    })
+  }
+
+  /**
+   * Return a mutation by name.
+   *
+   * @param {String} mutationName
+   */
+  getMutation (mutationName) {
+    return this.mutations.find(mutation => mutation.name === mutationName)
+  }
+
+  // --- GraphQL query subscriptions
+
+  /**
+   * @param {Vue} componentOrView
+   * @returns {Subscription}
+   */
+  getOrCreateSubscription (componentOrView) {
+    const queryName = componentOrView.query.name
+    let subscription = this.subscriptions[queryName]
+    if (!subscription) {
+      subscription = this.subscriptions[queryName] = new Subscription(componentOrView.query)
+    }
+    return subscription
+  }
+
+  /**
+   * @param {Vue} componentOrView
+   */
+  subscribe (componentOrView) {
+    // First we retrieve the existing, or create a new subscription (and add to the pool).
+    const subscription = this.getOrCreateSubscription(componentOrView)
+    if (!subscription.subscribers[componentOrView._uid]) {
+      // NOTE: make sure to remove it afterwards to avoid memory leaks!
+      subscription.subscribers[componentOrView._uid] = componentOrView
+      // Then we recompute the query, checking if variables match, and action name is set.
+      this.recompute(subscription)
+    }
+    // Otherwise we are calling subscribe for a component or view already subscribed.
+  }
+
+  /**
+   * Start any pending subscriptions.
+   */
+  startSubscriptions () {
+    const pendingSubscriptions = Object.values(this.subscriptions)
+      .filter(subscription => {
+        return subscription.observable === null || subscription.reload
+      })
+    pendingSubscriptions.forEach(subscription => this.startSubscription(subscription))
+  }
+
+  /**
+   * Start a Subscription for a view or component.
+   *
+   * @param {Subscription} subscription
+   */
+  startSubscription (subscription) {
+    if (this.debug) {
+      // eslint-disable-next-line no-console
+      console.debug(`Starting subscription ${subscription.query.name}`, subscription)
+    }
+    subscription.handleViewState(ViewState.LOADING, null)
+
+    // Stop if already running.
+    if (subscription.observable !== null) {
+      if (this.debug) {
+        // eslint-disable-next-line no-console
+        console.debug(`Subscription for query [${subscription.query.name}] already running. Stopping it...`)
+      }
+      this.stopSubscription(subscription)
+    }
+
+    try {
+      // Then start subscription.
+      subscription.observable = this.startDeltasSubscription(
+        subscription.query.query,
+        subscription.query.variables,
+        {
+          next: function next (response) {
+            subscription.actionNames.forEach(actionName => {
+              store.dispatch(actionName, response.data).then(() => {})
```

Comment:
Hmm, I wonder if I didn't speak too fast about fixing the issue of loading forever when there are no workflows.

Re-reading the code, now I can see that I'm changing the state immediately after I create the subscription, before it receives the first message where you pointed.

@oliver-sanders I can move the `subscription.handleViewState(ViewState.COMPLETE, null)` call here. The only downside I can see is that it would call the Vuex code for every subscription, which would be a bit expensive if you have a lot of GraphQL messages... :thinking:

Reply:
I'm not worried about moving the state change, was just wondering if something had been missed.

> Hmm, I wonder if I didn't speak too fast about fixing the issue of loading forever when there are no workflows.

Have re-tested, no probs.